kratos/geometries/nurbs_curve_on_surface_geometry.h

```diff
@@ -0,0 +1,294 @@
+    * @see PointLocalCoordinates
+    */
+    void GlobalSpaceDerivatives(
+        std::vector<CoordinatesArrayType>& rGlobalSpaceDerivatives,
+        const CoordinatesArrayType& rCoordinates,
+        const SizeType DerivativeOrder) const override
+    {
+        // Check size of output
+        if (rGlobalSpaceDerivatives.size() != DerivativeOrder + 1) {
+            rGlobalSpaceDerivatives.resize(DerivativeOrder + 1);
+        }
+
+        // Compute the gradients of the embedded curve in the parametric space of the surface
+        std::vector<array_1d<double, 3>> curve_derivatives;
+        mpNurbsCurve->GlobalSpaceDerivatives(curve_derivatives, rCoordinates, DerivativeOrder);
+        
+        // Compute the gradients of the surface in the geometric space
+        array_1d<double, 3> surface_coordinates =  ZeroVector(3);
+        surface_coordinates[0] = curve_derivatives[0][0];
+        surface_coordinates[1] = curve_derivatives[0][1];
+        std::vector<array_1d<double, 3>> surface_derivatives;
+        mpNurbsSurface->GlobalSpaceDerivatives(surface_derivatives, surface_coordinates, DerivativeOrder);
+
+        std::function<array_1d<double, 3>(int, int, int)> c;
+        c = [&](int DerivativeOrder, int i, int j) -> array_1d<double, 3> {
+            if (DerivativeOrder > 0) {
+                array_1d<double, 3> result = ZeroVector(3);
+
+                for (int a = 1; a <= DerivativeOrder; a++) {
+                    result += (
+                        c(DerivativeOrder - a, i + 1, j) * curve_derivatives[a][0] +
+                        c(DerivativeOrder - a, i, j + 1) * curve_derivatives[a][1]
+                        ) * NurbsUtilities::GetBinomCoefficient(DerivativeOrder - 1, a - 1);
+                }
+
+                return result;
+            }
+            else {
+                const int index = NurbsSurfaceShapeFunction::IndexOfShapeFunctionRow(i, j);
+                return surface_derivatives[index];
+            }
+        };
+        for (SizeType i = 0; i <= DerivativeOrder; i++) {
+            rGlobalSpaceDerivatives[i] = c(i, 0, 0);
+        }
+    }
+
+    ///@}
+    ///@name Information
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        return "2 dimensional nurbs curve on 3D surface.";
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "2 dimensional nurbs curve on 3D surface.";
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const override
+    {
+    }
+
+    ///@}
+private:
+    ///@name Private Static Member Variables
+    ///@{
+
+    static const GeometryData msGeometryData;
+
+    static const GeometryDimension msGeometryDimension;
+
+    ///@}
+    ///@name Private Member Variables
+    ///@{
+
+    typename NurbsSurfaceType::Pointer mpNurbsSurface;
+    typename NurbsCurveType::Pointer mpNurbsCurve;
```

Comment:
does this have to be a shared ptr?
(@tteschemacher )

Reply:
Not necessarily, but in this case it does not really make a difference.

If there will be constructors that construct e.g. the curve then it needs to be a smart pointer.